Add API documentation

rudigiesler opened the pull request.

rudigiesler commented:
Generated docs: http://seed-auth-api.readthedocs.io/en/feature-issue-1-add-api-documentation/http_api.html

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1035 @@
+
+   :<json str token: The provided token.
+   :<json str password: The new password.
+   :code 200: The password was successfully reset.
```
If we aren't returning any result, maybe we could use `204` here.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1035 @@
+
+   Reset the users password using the provided token.
+
+   :<json str token: The provided token.
```
Vaguely remember us discussing a reason for putting the token in the body here, but wondering if it would be more consistent to include the token in the header.

rudigiesler replied:
I think for this we don't want to use a normal user token, but one that's specially generated for password resets? We could put it in the header, but I think it's slightly different to the normal token.

justinvdm replied:
Ah, nevermind comment, for a moment forgot this is the endpoint for unauthenticated users.

justinvdm commented:
General comment: should we maybe use a [`422`](http://www.restpatterns.org/HTTP_Status_Codes/422_-_Unprocessable_Entity) rather than a `400`?

justinvdm commented:
Other general comment: should we use [`204`](http://www.restpatterns.org/HTTP_Status_Codes/204_-_No_Content) for empty success responses?

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1035 @@
+
+       {"name":"Brotherhood Without Banners","id":4,"teams":[],"url":"https://example.org/organizations/4","users":[]}
+
+.. http:delete:: /organizations/(int:organization_id)
```
Wondering if deleting organisations is something we want to support. Might result in lots of orphan data, or lots of related data being removed in the process. Maybe for now we should just not support it, and later consider supporting archiving (so they still exist, but have an `is_archived` flag we can filter on). Thoughts?

rudigiesler replied:
From the discussion:

For now, `DELETE` archives an organization, it doesn't delete it. It also archives all of that organization's teams.

You cannot add permissions to users, only to teams (this solves the problem of not removing relevant individual permissions when an organization is archived)

rudigiesler commented:
We can look at using `422` instead of `400`. I think by default django rest framework uses 400, but we can look at how difficult that will be to change to 422.

I'll change all empty responses to be `204`.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1088 @@
+
+   :<json str email: The email of the user to reset the password for.
+   :<json str app:
+        The application that the token should go to, configured in settings.
```
Should we allow this to be optional (in which case it goes to a default app)?

rudigiesler replied:
Yes, that was my idea behind it. I'll make the docs a bit clearer.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1092 @@
+
+        HTTP/1.1 204 No Content
+
+.. http:put:: /teams/(int:team_id)/permission
```
Should we use `POST /teams/<team_id>/permissions/` here for consistency?

rudigiesler replied:
Yes that's more correct, since we're creating a permission, not adding an existing permission to a team. I'll update it.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1093 @@
+        The application that the token should go to, configured in settings.
+        This value is optional, defaults to the default configured application.
+   :code 202: The password reset process was started.
+   :code 400: The username does not exist.
```
Should we maybe return `202` regardless to not leak information about registered users?

rudigiesler replied:
That sound good.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1094 @@
+
+      HTTP/1.1 204 No Content
+
+.. http:put:: /organizations/(int:organization_id)/user
```
Should we use `POST /organizations/<org_id>/user/`?

rudigiesler replied:
Wouldn't it be `/users/`, and I was not sure about a `POST`, since you're updating an organization by adding an existing user, not creating a new user.

justinvdm replied:
Woops, sorry, tyop, `users/` not `user/`.

I think `POST` makes sense, since you are adding a new resource to the `/organizations/<org_id>/users/` collection.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1099 @@
+
+        HTTP/1.1 204 No Content
+
+.. http:put:: /users/(int:user_id)/team
```
Similar question about using `POST .../teams/`

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1099 @@
+            "organizations": []
+        }
+
+.. http:put:: /users/(int:user_id)/organization
```
Is this endpoint needed since we have `/organizations/<org_id>/users/`?

rudigiesler replied:
I'm not sure? I think there are some places where the same functionality is replicated (due to it being a many to many link). Should we remove one? Which one?

justinvdm replied:
I vote we remove this one, I think we would only need `/organizations/<org_id>/users/` for adding users to organisations.

rudigiesler replied:
Okay sure. What about adding and removing users from teams? Should that be done on the user endpoints or the team endpoints?

justinvdm replied:
Woops, didn't catch that. +1 on `/teams/<team_id>/users/`.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1097 @@
+    :query string permission_contains:
+        All the permission fields on the resulting teams must contain this string.
+    :query string object_id:
+        All the object_id fields on the resulting teams must equal this string.
```
Do we need this query param if `permission_contains` allows us to filter on a permission's object id? Or is `permission_contains` for the permission `type` field only?

rudigiesler replied:
`permission_contain` is only for the `type` field. I didn't rename it when I renamed the permission things. It should be `type_contains`. I'll change it.

justinvdm replied:
Ah ok cool. A little bikesheddy, but I'd opt for us calling it `type` instead to be more consistent, and then just document that it accepts wildcards. Happy with `type_contains` though.

justinvdm commented:
Looks good :) I think the only unresolved thing is the to use `/teams/<team_id>/users/` instead of `/users/<user_id>/teams/`. Otherwise :+1:

rudigiesler commented:
@justinvdm I've changed `/users/<user_id>/teams/` to `/teams/<team_id>/users`, and changed to using token auth instead of JWT token auth. Ready for another review.

justinvdm commented on docs/http_api.rst:
```diff
@@ -0,0 +1,1064 @@
+user:create
+    Can create new users.
+
+.. http:get:: /permissions
```
Should this maybe be `/permissions/` for consistency?

justinvdm commented:
:+1:

Is the plan to update the `400`s in the docs to `422`s once know this is easy enough with django-rest-framework?

rudigiesler commented:
Yes. I'll have a look if it's easy enough to do. But if it involves adding a bunch of try catches and handling error cases myself, I might leave it as `400`s.

Actually, let me update things to `422`s, and we can switch them back in the docs if it turns out to be difficult. That way we won't forget about it.

rudigiesler commented:
@justinvdm Okay changes some of the `400`s to `401`s, and some to `422`s. Ready for one last look.

justinvdm commented:
:+1: looks awesome :)
